rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs

```diff
@@ -0,0 +1,467 @@
+                .async_sleep
+                .sleep(this.minimum_throughput.per_time_elapsed());
+            // If we already pushed a log during this poll, don't set a
+            // "floating back". Otherwise, do set one.
+            //
+            // The floating back acts as a end-bound for the purposes of
+            // calculating throughput for a given time span. 0-byte logs
+            // don't affect the throughput calculation except when they are
+            // the first or last log. The starting log is pushed the first
+            // time this is polled. If no data was emitted during a poll,
+            // then we set the floating back.
+            if !a_log_was_pushed_this_poll {
+                this.throughput_logs.set_floating_back(now);
+            }
+            // We also schedule a wake up for current task to ensure that
+            // it gets polled at least one more time.
+            cx.waker().wake_by_ref();
+        };
+        this.sleep_fut.replace(sleep_fut);
+
+        // Calculate the current throughput and emit an error if it's too low.
+        let actual_throughput = this.throughput_logs.calculate_throughput()?;
+        let is_below_minimum_throughput = actual_throughput
+            .map(|t| t < self.minimum_throughput)
+            .unwrap_or_default();
+        if is_below_minimum_throughput {
+            Poll::Ready(Some(Err(Box::new(Error::ThroughputBelowMinimum {
+                expected: self.minimum_throughput,
+                actual: actual_throughput.unwrap(),
+            }))))
+        } else {
+            poll_res
+        }
+    }
+
+    fn poll_trailers(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<http::HeaderMap>, Self::Error>> {
+        let now = self.time_source.now();
+        let mut a_log_was_pushed_this_poll = false;
+        // Attempt to read the data from the inner body, then update the
+        // throughput logs.
+        let mut this = self.as_mut().project();
+        let poll_res = match this.inner.poll_trailers(cx) {
+            Poll::Ready(Ok(Some(trailers))) => {
+                let size_of_all_keys: u64 = trailers.keys().map(|k| k.as_str().len() as u64).sum();
+                let size_of_all_values: u64 = trailers.values().map(|v| v.len() as u64).sum();
+                this.throughput_logs
+                    .push((now, size_of_all_keys + size_of_all_values));
+                a_log_was_pushed_this_poll = true;
+                Poll::Ready(Ok(Some(trailers)))
+            }
+            Poll::Pending => Poll::Pending,
+            // If we've read all the data or an error occurred, then return that result.
+            res => return res,
+        };
+
+        // Check the sleep future to see if it needs refreshing.
+        let mut sleep_fut = this.sleep_fut.take().unwrap_or_else(|| {
+            this.async_sleep
+                .sleep(this.minimum_throughput.per_time_elapsed())
```

Comment:
this can be simplified—poll_trailers can only be called once. The only thing we need to do here is watch for a timeout, we don't actually need to calculate the size of the trailers. I wonder if we actually need _anything_ in this method at all—once the last read of the body is complete, we essentially want to schedule a timeout for how long it's OK to wait for the trailers to be polled.